uportal-war/src/main/java/org/jasig/portal/jgroups/protocols/DAO_PING.java

```diff
@@ -142,8 +151,12 @@ protected void handleView(View view) {
         
         final Collection<Address> mbrs = view.getMembers();
         final boolean is_coordinator = !mbrs.isEmpty() && mbrs.iterator().next().equals(local_addr);
-        if (is_coordinator) {
+        if (is_coordinator != is_coord) {
+            log.warn("Inconsistent state: is_cood != calculated is_coordinator");
+        }
+        if (is_coordinator || is_coord) {
```

Comment:
Per jGroups standards the coordinator is the first node in the view.  See http://www.jgroups.org/manual/html/user-channel.html#JoiningACluster.

I don't think is_coordinator will ever be != is_coord.  is_coord is computed by parent class Discovery.java as
{code}
            case Event.VIEW_CHANGE:
                List<Address> tmp;
                view=(View)evt.getArg();
                if((tmp=view.getMembers()) != null) {
                    synchronized(members) {
                        members.clear();
                        members.addAll(tmp);
                    }
                }
                current_coord=!members.isEmpty()? members.get(0) : null;
                is_coord=current_coord != null && local_addr != null && current_coord.equals(local_addr);

                return down_prot.down(evt);
{code}
You see is_coord is calculated to be the first member in the view in the event being equal to the local_addr.  I believe calling down_prot.down(evt) eventually calls DAO_PING.down(event) which calls this method which does the identical computation.

I think the previous code was fine as far as it went.  I believe the issue we are running into at Texas A&M University (the UP_JGROUPS_PING table not getting cleared by the coordinator) is due to a combination of
* the coordinator crashing with an OutOfMemory error or other catastrophic event
* uPortal's implementation uses a database table to coordinate the members of the cluster rather than using Multicast messaging or a GossipRouter (see class documentation of http://jgroups.org/javadoc/org/jgroups/protocols/Discovery.html), the first of which and possibly the second also  would allow the remaining members of the cluster to realize the coordinator was dead.

I think we need to attack the core issue ... giving one or more nodes in the cluster the opportunity to realize the coordinator is dead and isn't updating the UP_JGROUPS_PING table.  

One approach to solve that issue is to add another column called 'last_updated' to UP_JGROUPS_PING table.  Each node in the cluster every DAO_PING.interval milliseconds posts its address to the table.  This should also update the timestamp stored in the last_updated column.  Then when a view change occurs if a node is not the coordinator have some mechanism where it checks the coodinator's entry in the UP_JGROUPS_PING table and if it is older than interval*2, continue through and act like the coordinator to update the table, possibly insuring the coordinator's entry is not in the view.  I'm not sure if this strategy will cause a loop or race condition where multiple nodes try to sync the table or if whomever is the first remaining node in the view doesn't know they are now the coordinator (would we need to trigger a view change event to notify everyone the coordinator is down and would it get broadcast to the nodes in the cluster?).  We should look at any recovery strategy carefully to insure it doesn't cause cycles or still have an issue.  In development setting a breakpoint in the JVM should freeze it enough to allow us to test the recovery mechanism reasonably easily.